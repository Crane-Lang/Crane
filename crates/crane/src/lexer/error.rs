use serde::{Serialize, Deserialize};
use thiserror::Error;

use crate::ast::Span;

#[derive(Error, Debug, PartialEq, Clone, Serialize, Deserialize)]
pub struct LexError {
    pub kind: LexErrorKind,
    pub span: Span,
}

impl Default for LexError {
    fn default() -> Self {
        Self {
            kind: Default::default(),
            span: Span::new(0, 0),
        }
    }
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.kind)
    }
}

#[derive(Debug, Error, PartialEq, Clone, Default, Serialize, Deserialize)]
pub enum LexErrorKind {
    #[default]
    #[error("unknown")]
    Unknown,
}
