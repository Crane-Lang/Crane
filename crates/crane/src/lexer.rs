mod error;

pub mod token;

pub use error::*;

use logos::Logos;

use crate::lexer::token::{Token, TokenKind};

pub struct Lexer<'src> {
    lexer: logos::Lexer<'src, TokenKind>,
}

impl<'src> Lexer<'src> {
    pub fn new(input: &'src str) -> Self {
        Self {
            lexer: TokenKind::lexer(input),
        }
    }
}

impl<'src> Iterator for Lexer<'src> {
    type Item = Result<Token, LexError>;

    fn next(&mut self) -> Option<Self::Item> {
        let kind = self.lexer.next()?;
        let lexeme = self.lexer.slice();
        let span = self.lexer.span().into();

        Some(
            kind.map(|kind| Token {
                kind,
                lexeme: lexeme.into(),
                span,
            })
            .map_err(|kind| LexError { kind, span }),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn lex(input: &str) -> logos::Lexer<TokenKind> {
        TokenKind::lexer(input)
    }

    fn check(input: &str, token: TokenKind) {
        let mut lexer = TokenKind::lexer(input);

        assert_eq!(lexer.next(), Some(Ok(token)));
        assert_eq!(lexer.slice(), input);
    }

    #[test]
    fn lex_open_paren() {
        check("(", TokenKind::OpenParen)
    }

    #[test]
    fn lex_close_paren() {
        check(")", TokenKind::CloseParen)
    }

    #[test]
    fn lex_open_brace() {
        check("{", TokenKind::OpenBrace)
    }

    #[test]
    fn lex_close_brace() {
        check("}", TokenKind::CloseBrace)
    }

    #[test]
    fn lex_comment() {
        check("// This is a comment.", TokenKind::Comment)
    }

    #[test]
    fn lex_fn_declaration() {
        let mut lex = lex("fn foo() {}");

        assert_eq!(lex.next(), Some(Ok(TokenKind::Ident)));
        assert_eq!(lex.span(), 0..2);
        assert_eq!(lex.slice(), "fn");

        assert_eq!(lex.next(), Some(Ok(TokenKind::Ident)));
        assert_eq!(lex.span(), 3..6);
        assert_eq!(lex.slice(), "foo");

        assert_eq!(lex.next(), Some(Ok(TokenKind::OpenParen)));
        assert_eq!(lex.span(), 6..7);
        assert_eq!(lex.slice(), "(");

        assert_eq!(lex.next(), Some(Ok(TokenKind::CloseParen)));
        assert_eq!(lex.span(), 7..8);
        assert_eq!(lex.slice(), ")");

        assert_eq!(lex.next(), Some(Ok(TokenKind::OpenBrace)));
        assert_eq!(lex.span(), 9..10);
        assert_eq!(lex.slice(), "{");

        assert_eq!(lex.next(), Some(Ok(TokenKind::CloseBrace)));
        assert_eq!(lex.span(), 10..11);
        assert_eq!(lex.slice(), "}");
    }

    #[test]
    fn test_lexer() {
        insta::glob!("snapshot_inputs/*.crane", |path| {
            let source = std::fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&source);

            insta::assert_yaml_snapshot!(lexer.into_iter().collect::<Vec<_>>());
        })
    }
}
