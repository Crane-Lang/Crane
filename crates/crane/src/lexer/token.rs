use logos::Logos;
use smol_str::SmolStr;

use crate::ast::Span;
use crate::lexer::LexErrorKind;

#[derive(Logos, Debug, PartialEq, Clone, Copy)]
#[logos(error = LexErrorKind)]
pub enum TokenKind {
    /// `(`
    #[token("(")]
    OpenParen,

    /// `)`
    #[token(")")]
    CloseParen,

    /// `{`
    #[token("{")]
    OpenBrace,

    /// `}`
    #[token("}")]
    CloseBrace,

    /// `,`
    #[token(",")]
    Comma,

    /// An identifier.
    #[regex(r"[A-Za-z_]+")]
    Ident,

    /// A string literal.
    #[regex(r#""[^"]*""#)]
    String,

    /// An integer literal.
    #[regex(r#"[\d]+"#)]
    Integer,

    /// A comment.
    #[regex(r"//.*")]
    Comment,

    /// Any sequence of whitespace characters.
    #[regex(r"[ \n]+", logos::skip)]
    Whitespace,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub lexeme: SmolStr,
    pub span: Span,
}
