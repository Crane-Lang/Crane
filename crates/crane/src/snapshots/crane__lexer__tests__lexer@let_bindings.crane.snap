---
source: crates/crane/src/lexer.rs
expression: "lexer.into_iter().collect::<Vec<_>>()"
input_file: crates/crane/src/snapshot_inputs/let_bindings.crane
---
- Ok:
    kind: Ident
    lexeme: use
    span:
      start: 0
      end: 3
- Ok:
    kind: Ident
    lexeme: std
    span:
      start: 4
      end: 7
- Ok:
    kind: ColonColon
    lexeme: "::"
    span:
      start: 7
      end: 9
- Ok:
    kind: Ident
    lexeme: int
    span:
      start: 9
      end: 12
- Ok:
    kind: ColonColon
    lexeme: "::"
    span:
      start: 12
      end: 14
- Ok:
    kind: Ident
    lexeme: int_to_string
    span:
      start: 14
      end: 27
- Ok:
    kind: Ident
    lexeme: use
    span:
      start: 28
      end: 31
- Ok:
    kind: Ident
    lexeme: std
    span:
      start: 32
      end: 35
- Ok:
    kind: ColonColon
    lexeme: "::"
    span:
      start: 35
      end: 37
- Ok:
    kind: Ident
    lexeme: io
    span:
      start: 37
      end: 39
- Ok:
    kind: ColonColon
    lexeme: "::"
    span:
      start: 39
      end: 41
- Ok:
    kind: Ident
    lexeme: print
    span:
      start: 41
      end: 46
- Ok:
    kind: Ident
    lexeme: use
    span:
      start: 47
      end: 50
- Ok:
    kind: Ident
    lexeme: std
    span:
      start: 51
      end: 54
- Ok:
    kind: ColonColon
    lexeme: "::"
    span:
      start: 54
      end: 56
- Ok:
    kind: Ident
    lexeme: io
    span:
      start: 56
      end: 58
- Ok:
    kind: ColonColon
    lexeme: "::"
    span:
      start: 58
      end: 60
- Ok:
    kind: Ident
    lexeme: println
    span:
      start: 60
      end: 67
- Ok:
    kind: Ident
    lexeme: pub
    span:
      start: 69
      end: 72
- Ok:
    kind: Ident
    lexeme: fn
    span:
      start: 73
      end: 75
- Ok:
    kind: Ident
    lexeme: main
    span:
      start: 76
      end: 80
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 80
      end: 81
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 81
      end: 82
- Ok:
    kind: OpenBrace
    lexeme: "{"
    span:
      start: 83
      end: 84
- Ok:
    kind: Ident
    lexeme: let
    span:
      start: 89
      end: 92
- Ok:
    kind: Ident
    lexeme: name
    span:
      start: 93
      end: 97
- Ok:
    kind: Equal
    lexeme: "="
    span:
      start: 98
      end: 99
- Ok:
    kind: String
    lexeme: "\"Arya\""
    span:
      start: 100
      end: 106
- Ok:
    kind: Ident
    lexeme: let
    span:
      start: 111
      end: 114
- Ok:
    kind: Ident
    lexeme: gold
    span:
      start: 115
      end: 119
- Ok:
    kind: Equal
    lexeme: "="
    span:
      start: 120
      end: 121
- Ok:
    kind: Integer
    lexeme: "100"
    span:
      start: 122
      end: 125
- Ok:
    kind: Ident
    lexeme: print
    span:
      start: 131
      end: 136
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 136
      end: 137
- Ok:
    kind: String
    lexeme: "\"Hello, \""
    span:
      start: 137
      end: 146
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 146
      end: 147
- Ok:
    kind: Ident
    lexeme: print
    span:
      start: 152
      end: 157
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 157
      end: 158
- Ok:
    kind: Ident
    lexeme: name
    span:
      start: 158
      end: 162
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 162
      end: 163
- Ok:
    kind: Ident
    lexeme: println
    span:
      start: 168
      end: 175
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 175
      end: 176
- Ok:
    kind: String
    lexeme: "\".\""
    span:
      start: 176
      end: 179
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 179
      end: 180
- Ok:
    kind: Ident
    lexeme: print
    span:
      start: 186
      end: 191
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 191
      end: 192
- Ok:
    kind: String
    lexeme: "\"You currently have \""
    span:
      start: 192
      end: 213
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 213
      end: 214
- Ok:
    kind: Ident
    lexeme: print
    span:
      start: 219
      end: 224
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 224
      end: 225
- Ok:
    kind: Ident
    lexeme: int_to_string
    span:
      start: 225
      end: 238
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 238
      end: 239
- Ok:
    kind: Ident
    lexeme: gold
    span:
      start: 239
      end: 243
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 243
      end: 244
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 244
      end: 245
- Ok:
    kind: Ident
    lexeme: println
    span:
      start: 250
      end: 257
- Ok:
    kind: OpenParen
    lexeme: (
    span:
      start: 257
      end: 258
- Ok:
    kind: String
    lexeme: "\" gold at your disposal.\""
    span:
      start: 258
      end: 283
- Ok:
    kind: CloseParen
    lexeme: )
    span:
      start: 283
      end: 284
- Ok:
    kind: CloseBrace
    lexeme: "}"
    span:
      start: 285
      end: 286

